contracts/chain-adapters/ForwarderBase.sol

```diff
@@ -0,0 +1,216 @@
+     */
+    function __Forwarder_init(address _crossDomainAdmin) public onlyInitializing {
+        __UUPSUpgradeable_init();
+        _setCrossDomainAdmin(_crossDomainAdmin);
+    }
+
+    /**
+     * @notice Sets a new cross domain admin for this contract.
+     * @param _newCrossDomainAdmin L1 address of the new cross domain admin.
+     */
+    function setCrossDomainAdmin(address _newCrossDomainAdmin) external onlyAdmin {
+        _setCrossDomainAdmin(_newCrossDomainAdmin);
+    }
+
+    /**
+     * @notice Sets a new RouteIndex to a specified address.
+     * @param _destinationAddress The address to set in the availableRoutes mapping.
+     * @param _routeIndex RouteIndex struct containing the next immediate path to reach the _destination address.
+     * @dev Each forwarder will not know how many layers it must traverse to arrive at _destinationAddress until it is precisely
+     * one layer away. The availableRoutes mapping lets the forwarder know what path it should take to progess, and then relies on
+     * the forwarders on the subsequent layers to derive the next route in sequence.
+     */
+    function updateRoute(address _destinationAddress, RouteIndex memory _routeIndex) external onlyAdmin {
+        availableRoutes[_destinationAddress] = _routeIndex;
+        emit RouteUpdated(_destinationAddress, _routeIndex);
+    }
+
+    /**
+     * @notice Sets a new remote token in the remoteTokens mapping.
+     * @param _destinationAddress The address to set in the remoteTokens mapping. This is to identify the network associated with the
+     * remote token.
+     * @param _baseToken The address of the token which exists on the current network.
+     * @param _remoteToken The address of the token on the network which is next on the path to _destinationAddress.
+     * @dev This mapping also relies on using _destinationAddress to determine the network to bridge to.
+     */
+    function updateRemoteToken(
+        address _destinationAddress,
+        address _baseToken,
+        address _remoteToken
+    ) external onlyAdmin {
+        remoteTokens[_destinationAddress][_baseToken] = _remoteToken;
+        emit RemoteTokensUpdated(_destinationAddress, _baseToken, _remoteToken);
+    }
+
+    /**
+     * @notice Relays a specified message to a target on an arbitrary layer.
+     * @param target The address of the contract that will receive the input message.
+     * @param message The data to execute on the contract.
+     * @dev Each forwarder will only know if they are on the layer directly before the layer which contains the `target` contract.
+     * If the next layer contains the target, we send `message` cross-chain normally. Otherwise, we wrap `message` into a `relayMessage`
+     * call to the next layer's forwarder contract.
+     */
+    function relayMessage(address target, bytes memory message) external payable override onlyAdmin {
+        RouteIndex storage route = availableRoutes[target];
```

Comment:
I know we are constrained here but its really confusing that we grab information about the next layer to forward the message to using the `target` as the key---and then we subsequently check if the `target` is equal to any of hte values in the route information.

- `route` is an inaccurate name. We are loading information of known contracts we could target on the next layer. We are comparing the `target`--the end target of the `message`--with contracts from `availableRoutes[target]`.
- `availableRoutes`: this is really list of possible target contracts. Could it be renamed `destinationChainTargets`?
- I think `RouteIndex` can be better named. Perhaps `PossibleTargetContracts` or `TargetContracts`
- See comment below about `remoteTokens` but a consistent name with these suggestions would be `destinationChainTokens`